readhomer_atlas/library/importers.py

```diff
@@ -1,70 +1,111 @@
 import json
 import os
+import re
+import sys
+from collections import defaultdict
 
 from django.conf import settings
 
-from .models import Book, Line, Version
+from .models import Node
 
 
 LIBRARY_DATA_PATH = os.path.join(settings.PROJECT_ROOT, "data", "library")
 LIBRARY_METADATA_PATH = os.path.join(LIBRARY_DATA_PATH, "metadata.json")
 
 
-def _prepare_line_obj(version_obj, book_lookup, counters, line, line_idx):
-    ref, tokens = line.strip().split(maxsplit=1)
-    _, textpart_ref = ref.split(".", maxsplit=1)
-    book_ref, line_ref = textpart_ref.split(".", maxsplit=1)
+class CTSImporter:
+    """
+    urn:cts:CTSNAMESPACE:WORK:PASSAGE
+    https://cite-architecture.github.io/ctsurn_spec
+    """
 
-    book_obj = book_lookup.get(book_ref)
-    if book_obj is None:
-        book_obj, _ = Book.objects.get_or_create(
-            version=version_obj,
-            position=int(book_ref),
-            idx=counters["book_idx"],
-            ref=book_ref,
-            urn=Book.generate_urn(version_obj.urn, book_ref),
+    CTS_URN_SCHEME = ["nid", "namespace", "textgroup", "work", "version"]
+
+    def __init__(self, version_data, nodes=dict()):
+        self.version_data = version_data
+        self.nodes = nodes
+        self.urn = self.version_data["urn"].strip()
+        self.metadata = self.version_data["metadata"]
+        self.citation_scheme = self.metadata["citation_scheme"]
+        self.name = self.metadata["work_title"]
+        self.idx_lookup = defaultdict(int)
+        self.root = None
+
+    def get_node_idx(self, kind):
+        idx = self.idx_lookup[kind]
+        self.idx_lookup[kind] += 1
+        return idx
+
+    def urn_has_exemplar(self, node_urn):
+        return len(node_urn.rsplit(":")[-2].split(".")) == 4
+
+    def get_urn_scheme(self, node_urn):
+        if self.urn_has_exemplar(node_urn):
+            return [*self.CTS_URN_SCHEME, "exemplar", *self.citation_scheme]
+        return [*self.CTS_URN_SCHEME, *self.citation_scheme]
+
+    def destructure_node(self, node_urn, tokens):
+        split = ["urn:cts", ":", *re.split(r"([:|.])", node_urn)[4:]]
```

Comment:
I think we'll end up being able to leverage ![](https://github.trello.services/images/mini-trello-icon.png) [Build or re-use a Python library for URN manipulationi](https://trello.com/c/rzaNKWCi) in the future; this is fine for now.